subsys/net/lib/async_socket/async_socket.c

```diff
@@ -0,0 +1,380 @@
+		rcv_cb->buf = NULL;
+		rcv_cb->max_len = 0;
+		rcv_cb->cb = NULL;
+		rcv_cb->cb_data = NULL;
+
+		num_rcv_cbs--;
+
+		/* Now, signal the signal_sock to restart the poll server,
+		 * so it can recreate the ufds[] list and start poll()
+		 * *without* the newly un-registered socket.
+		 */
+		status = async_server_restart();
+	}
+}
+
+/* Fill the poll() ufds argument from the rcv_callback list: */
+static void rcv_callbacks_to_ufds(struct pollfd *fds)
+{
+	struct rcv_callbacks *entry, *end;
+	int fds_index = 0;
+
+	end = rcv_callback + ARRAY_SIZE(rcv_callback);
+	for (entry = &rcv_callback[0], fds_index = 0; entry < end; entry++) {
+		if (entry->sock != INVALID_SOCK) {
+			fds[fds_index].fd = entry->sock;
+			fds[fds_index].events = POLLIN;
+			fds_index++;
+		}
+	}
+}
+
+/*
+ * Call poll() in a loop, waiting for registered async sockets
+ * to get signalled; if so, call their registered receive callbacks.
+ *
+ * Currently, this only handles receive.  Send, accept and other
+ * events are not handled.
+ *
+ * Problem: how to break out of a blocking poll() to reset
+ * the list of events?	(Without using a finite timeout -
+ * which is not good from a power perspective).
+ * In Linux, we could use eventfd() to add an fd to the poll()
+ * list of fds, to enable breaking out of the poll() when
+ * a new fd is added/removed.
+ * But, Zephyr's poll() API only supports sockets for now, and Zephyr
+ * does not yet support something like eventfd().
+ * Calling close() on a bespoke socket placed in the list just
+ * for this purpose *might* unblock a
+ * poll()/select() operation with a POLLHUP event,
+ * but that is not implemented, and not clear in the POSIX standard.
+ *
+ * Solution: Reserve a UDP socket to 'signal' by sending a
+ * small message to the LOOPBACK address, unblocking the poll.
+ * The loopback is processed in Zephyr IP stack before calling
+ * the L2 driver, so this should work generally.
+ */
+void async_sock_server(void *unused1, void *unused2, void *unused3)
+{
+	ARG_UNUSED(unused1);
+	ARG_UNUSED(unused2);
+	ARG_UNUSED(unused3);
+
+	static struct pollfd ufds[MAX_RCV_CALLBACKS];
+	static struct sockaddr_storage from;
+	int nfds;
+	struct pollfd *entry, *end;
+	struct rcv_callbacks *cb_entry;
+	size_t size;
+	int rcv_len;
+	size_t from_len;
+
+	/* signal_sock is never closed, and always remains in first position */
+	ufds[0].fd = signal_sock;
+	ufds[0].events = POLLIN;
+
+	/* End of ufds[] array */
+	end = ufds + MAX_RCV_CALLBACKS;
+
+	while (1) {
+		/* Update array of polling structs: */
+		rcv_callbacks_to_ufds(&ufds[1]);
+
+		/* Wait until any socket gets signalled: */
+		nfds = poll(ufds, (num_rcv_cbs + 1), K_FOREVER);
+		NET_ASSERT(nfds != 0);  /* Timeout should be impossible */
+
+		if (nfds > 0) {
+			/* signal_sock signalled via loopback msg: */
+			if (ufds[0].revents & POLLIN) {
+				NET_ASSERT(ufds[0].fd == signal_sock);
+
+				/* Just get and discard data: */
+				rcv_len = recvfrom(ufds[0].fd, server_rcv_buf,
```

Comment:
This is another thing how this "async socket" would differ from a generic event loop: such an event loop doesn't necessary read data for an app, why - it can to it itself. (Depends, for a truly async, non-blocking sockets, handling reads/writes may be a big convenience).